spring-kafka/src/main/java/org/springframework/kafka/core/DefaultKafkaProducerFactory.java

```diff
@@ -1159,22 +1194,30 @@ public void close(@Nullable Duration timeout) {
 					this.removeProducer.test(this, this.producerFailed instanceof TimeoutException
 							? CLOSE_TIMEOUT_AFTER_TX_TIMEOUT
 							: timeout);
+					this.delegate.close(timeout == null ? this.closeTimeout : this.producerFailed instanceof TimeoutException
+							? CLOSE_TIMEOUT_AFTER_TX_TIMEOUT
+							: timeout);
 				}
 				else {
 					this.closed = this.removeProducer.test(this, timeout);
+					if (this.closed) {
+						this.delegate.close(timeout == null ? this.closeTimeout : timeout);
+					}
 				}
 			}
 		}
 
-		void closeDelegate(Duration timeout, List<Listener<K, V>> listeners) {
+		void closeDelegate(Duration timeout) {
```

Comment:
I modified the behavior of `CloseSafeProducer` to enhance the consistency of the close operation. Now, both `close()` and `closeDelegate()` methods invoke `removeProducer()`, ensuring that external logic (e.g., cache returning or listeners) behaves consistently for both operations. The key distinction is that `closeDelegate()` explicitly forces the closure of the producer. I believe this change makes the code more consistent and clearer. @artembilan, WDYT?